Increase spacing scale

gladwearefriends opened the pull request:
TLDR

With `primer-marketing` we now have an extended spacing scale to use for y-axis spacing 
```
$spacer-7:  48px
$spacer-8:  64px
$spacer-9:  80px
$spacer-10: 96px
$spacer-11: 112px
$spacer-12: 128px

and corresponding utility spacers for
mt-, mb-, my-, pt-, pb-, py-

e.g. mt-7, mb-9, py-11
```

Testing on github/github here: https://github.com/github/github/pull/81032
Deployed to review lab here: https://testing-primer.review-lab.github.com/
To test on review-lab, slap some of those new classes onto some elements!

This has also been tested in storybook.

---

Hi friends!
Wanted to get a PR open to see if this works and also so we could discuss about exactly which increments to increase our spacing scale with. Previous conversation at https://github.com/github/design-systems/issues/378 where we established we needed this on marketing sites.

Some ideas that have been proposed.
After 40px,
- jump by 16px
- jump by 40px
- jump by powers of 2 (64, 128, 256)

 **In this PR i went ahead and jumped it by 16px. Only adding spacer utilities for y-directional spacing. i.e. mt-7, pb-7, my-7**
- jump by 16px (56, 72, 88, 104, 120)
  - PROS: addresses the various spacers we have now with the `.featurette` (50, 60, 80) and `.jumbotron` (60, 100, 120) classes for github marketing pages.
  - CONS: Bloats up our margin and padding utilities a bit. **But ive gone ahead and put this into only our primer-marketing-support and primer-marketing-utilities modules** Not sure if how i did this was the most elegant though. We need to run the new `$spacers` variable through the margin & padding utility mixins again but not the most DRY way of doing that.
- jump by 40px
  - PROS: addresses most of our needs for spacers without bloating our utilities
  - CONS: doesnt address all of our current spacers that we use. jumping from 40 to 80, skips out on the 50/60px range we currently achieve with the `.featurette` and `.jumbotron` classes
- jump by powers of 2 (64, 128, 256)
  - PROS: gets big quickly, no bloat
  - CONS: gets big quickly 🙃 not as useful in some cases and arguable if we really need a 256 spacer

/cc @primer/ds-core @sophshep

gladwearefriends commented:
i'm not sure how to fix this error

```Error: index out of bounds for `nth($list, $n)`
 at Object.module.exports.renderSync (/home/travis/build/primer/primer-css/node_modules/primer-module-build/node_modules/node-sass/lib/index.js:434:16)```

i double checked my usage of `nth($list, $n)` a few times already 🤔

sophshep commented on modules/primer-marketing-utilities/lib/margin.scss:
```diff
@@ -0,0 +1,80 @@
+// Margin spacer utilities
+// stylelint-disable block-opening-brace-space-before, declaration-colon-space-before, primer/selector-no-utility, comment-empty-line-before
+@for $i from 1 through length($spacers) {
```
Likewise here, we don't need this full loop since it will generate CSS for allll the spacer variables instead of just the biggos you're adding.

Two options:

* Specify which values you want to run the loop to limit it to just the new marketing spacers
* Ditch the fancy Sass loop all together and write it manually

gladwearefriends replied:
i'm gonna specify the new variables so the loop only goes through the new marketing spacers

sophshep commented on modules/primer-marketing-utilities/lib/margin.scss:
```diff
@@ -0,0 +1,80 @@
+  /* Set a #{$size} margin on the top */
+  .mt-#{$scale} { margin-top   : #{$size} !important; }
+  /* Set a #{$size} margin on the right */
+  .mr-#{$scale} { margin-right : #{$size} !important; }
```
Do we need margin right and left for these? I can't think of any situations where I've wanted bigger values for these than what we've got. Currently, I feel like we just need the vertical values (to allow us to refactor jumbotron and featurette custom css). Taking these out will reduce a lot of final CSS.

(I'd be curious to hear from others on @primer/design-systems on if they agree)

gladwearefriends replied:
im gonna remove them unless others think otherwise! @primer/design-systems

sophshep commented on modules/primer-marketing-utilities/lib/margin.scss:
```diff
@@ -0,0 +1,80 @@
+  /* Set a #{$size} margin on the left */
+  .ml-#{$scale} { margin-left  : #{$size} !important; }
+  /* Set a negative #{$size} margin on top */
+  .mt-n#{$scale} { margin-top   : -#{$size} !important; }
```
I don't think we need biggo negative margins.

sophshep commented on modules/primer-marketing-utilities/lib/padding.scss:
```diff
@@ -0,0 +1,67 @@
+// Padding spacer utilities for marketing
+// stylelint-disable block-opening-brace-space-before, declaration-colon-space-before
+// stylelint-disable comment-empty-line-before
+@for $i from 1 through length($spacers) {
```
All the notes I left for margin apply to padding too.

sophshep commented on modules/primer-marketing-utilities/lib/margin.scss:
```diff
@@ -0,0 +1,41 @@
+// Margin spacer utilities for marketing
+// Utilities only added for y-direction margin (i.e. top & bottom)
+// stylelint-disable block-opening-brace-space-before, declaration-colon-space-before, primer/selector-no-utility, comment-empty-line-before
+@for $i from 1 through length($marketingSpacers) {
```
Niiice

gladwearefriends commented on modules/primer-marketing-support/lib/variables.scss:
```diff
@@ -1,3 +1,13 @@
 // Type
 $alt-body-font: Roboto, -apple-system, BlinkMacSystemFont, "Helvetica Neue", "Segoe UI", "Oxygen", "Ubuntu", "Cantarell", "Open Sans", sans-serif;
 $alt-mono-font: $mono-font;
```
hmm should I move these variables to the `primer-marketing-type` module? @sophshep @broccolini

broccolini commented on modules/primer-marketing-utilities/docs/margin.md:
```diff
@@ -0,0 +1,38 @@
+
+Use y-directional utilities to apply margin to the top, bottom or Y axis of an element. Directional utilities can change or override default margins, and can be used with a spacing scale of 7-11.
+
+Note: The marketing spacing scale extends the core 0-6 scale. See [primer-marketing-support](https://github.com/primer/primer-css/tree/master/modules/primer-marketing-support) for the extended marketing spacing scale.
```
You can remove this note if you update the intro para as per comment above.

broccolini commented on modules/primer-marketing-utilities/docs/margin.md:
```diff
@@ -0,0 +1,38 @@
+
+## Responsive y-directional margins
+
+All y-directional margin utilities can be adjusted per [breakpoint](/styleguide/css/modules/grid#breakpoints) using the following formula: `m[y-direction]-[breakpoint]-[spacer]`. Each responsive style is applied to the specified breakpoint and up.
```
Suggest remove `y-directional` with marketing:

> All marketing margin utilities can be adjusted per...

broccolini commented on modules/primer-marketing-utilities/docs/margin.md:
````diff
@@ -0,0 +1,38 @@
+</div>
+```
+
+## Responsive y-directional margins
````
Suggest changing heading to: `Responsive y-axis margin utilities`

broccolini commented on modules/primer-marketing-utilities/docs/padding.md:
```diff
@@ -0,0 +1,38 @@
+status_issue: https://github.com/github/design-systems/issues/378
+---
+
+Padding spacing utilities can be applied to elements to achieve bigger y-directional spaces beyond the `primer-core` spacing scale.
```
Could you make the same changes to padding based on the changes I suggested for margins?

broccolini commented on modules/primer-marketing-utilities/docs/margin.md:
```diff
@@ -0,0 +1,38 @@
+
+## Y-Directional margin spacing
+
+Use y-directional utilities to apply margin to the top, bottom or Y axis of an element. Directional utilities can change or override default margins, and can be used with a spacing scale of 7-11.
```
I notice you're using "y-directional" quite a lot through the docs. It doesn't really roll off the tongue very well and I don't think it's a commonly used term 😬. In this instance I think you can get away from using by changing the sentence to the following:

> Use marketing margin utilities to apply margin to top, bottom, or y-axis of an element.

I know it says "margin" twice but kinda difficult to get away from that here.

gladwearefriends replied:
Haha great point @broccolini ! I think i pulled it from our usage of "directional margins" and ran with it... agreed, axis is a much better term. I'll update it!

broccolini commented:
😬 😞  🙈 Going to pre-face this comment with a huge _sorry_ to @gladwearefriends for not thinking of this earlier, but while I was reviewing the pr, something dawned on me and I think best to share now rather than ignore it. I had looked at the initial values and thought _"they seem fine, if that's what works for marketing then cool"_, but then something started to bother me...

I kept looking at 7 to 11 thinking this feels like an odd range, why don't we have 6 like core? And then I thought about how it's bothered me for a while that we start with a half step of `4px` in core, and so I decided to map out the spacing scale to see how it would affect things if we stepped in increments of `8px` throughout core, and how that would then impact marketing.

This led me to seeing what happens if we started the marketing scale on 48 (because if we re-mapped core to `8px` steps from 1-6 we'd end on `48px` instead of `40px`. When I stepped the marketing scale from `48px` and up, the scale lands on a lot more values that map to the rest of the scales in core and marketing. It lands on a lot more powers of 2 which essentially make for more _composable_ numbers, i.e. more numbers that can be divided by or added up to make other numbers in our scales. 

This is what I aimed for with our other scales because design is often about horizontally and vertically lining a bunch of small and medium sized elements with other larger elements etc. Here's the re-mapped scale starting from `48px` which I quickly checked against the type and spacing scales:

| Spacer | Value | Is it highly composable? |
| --- | --- | --- |
| 7  | 48 | Medium: maps to type scale 00, 2, & 6; and spacer 10 |
| 8  | 64 | High: power of 2; maps to spacer 1,2,3, & 5; type scale 4 & 1 |
| 9  | 80 | Medium: double spacer 5, & type scale 3 and 0 |
| 10 | 96 | Medium: maps to type scale 00, 2, & 6; and spacer 7 |
| 11 | 112 | Low: maps to type scale 5 (body font size) |
| 12 | 128 | High: power of 2; maps to spacer 1,2,3, & 5; type scale 4 & 1 |

I think the marketing scale could have 6 steps like the core scale, that feels more natural and what people would expect. It also means starting the scale on a half step just like core. This is a fault of our core scale starting on a half step, we could have separated the `4px` value out and called it "nudge" or "half-step" or something, and then had a more logical scale that always stepped in increments of `8px`.

I know the timing of this is not good, it's very late in the process. We're working up to a release next week, and @gladwearefriends ❤️ ❤️ ❤️ has done a ton of work on adding the Sass for the new scale, writing docs, and updating the GitHub codebase. However, after having shipped a not-so-great spacing scale in the past and then having to update it (ht Rohan and Mu-an), it's a painful process and could be worth the effort now to ship this with a better more composable scale—assuming it actually works for marketing too.

@sophshep @gladwearefriends @jonrohan I'd like to hear what you think about this proposed scale? If you agree there's a number of ways we can go about rolling this out, and I also sincerely want to avoid extra pressure and work to @gladwearefriends - some options might be delaying v10, doing a minor release shortly after, and @primer/ds-core could take on the work to update too.

Of course another option is that you might feel this updated version doesn't provide you enough benefits and feel the current proposed scale works better - I'm open to that too.

sophshep commented:
The scale you're proposing looks good to me @broccolini. It will still give us all the values needed for marketing sites. You're right to speak up now vs. after it's shipped and is in use -- not hard to change a few values at this stage in the game!

gladwearefriends commented on modules/primer-marketing-support/README.md:
```diff
@@ -38,6 +27,24 @@ The source files included are written in [Sass][sass] (`scss`) You can simply po
+
+This is useful for when we want to achieve bigger vertical spacing between sections on marketing sites.
+
+See [primer-marketing-utilities](https://github.com/primer/primer-css/tree/master/modules/primer-marketing-utilities) for related spacing utilities.
```
good point! changing this to `[primer-marketing-utilities](../primer-marketing-utilities)`

gladwearefriends commented:
@broccolini i updated the scale and updated the docs to reflect the new scale as well as remove my poor usage of `y-directional` everywhere 😬 

thank you so much for reading everything carefully and putting thought into it. i feel like ive learned a lot about how our primer repo is set up and what things to consider when increasing the spacing scale (i.e. emphasis on composability of the values and what primer users would expect (marketing scale should mimic core scale))

gladwearefriends commented on modules/primer-marketing-support/README.md:
```diff
@@ -9,17 +9,6 @@
-
-<!-- %enddocs -->
-
 ## Install
```
Usage is already included below `## Install` here :) https://github.com/primer/primer/pull/376/files/7c8b3be162835cc4d51fedbb9f74d6f9036198ed#diff-e22f2520c00c79678dbb5d56c4e38b6aR20

broccolini replied:
Oh sorry I missed that!

broccolini commented:
@gladwearefriends This looks good to go! I'll update your github pr before merging this one in so we can test.

broccolini commented:
@gladwearefriends confirmed changes look good in github/github so I'm merging this in.
